test/distributed/org/apache/cassandra/distributed/upgrade/MixedModeTTLOverflowUpgradeTest.java

```diff
@@ -0,0 +1,169 @@
+                cluster.get(1).startup();
+                cluster.get(2).shutdown().get();
+                cluster.get(2).config().set("storage_compatibility_mode", NONE.toString());
+                cluster.get(2).startup();
+                assertPolicyTriggersAt2106(cluster.coordinator(1));
+                assertPolicyTriggersAt2106(cluster.coordinator(2));
+            }
+        });
+    }
+
+    @Test
+    public void testTTLOverflowAfterUpgrade() throws Throwable
+    {
+        testTTLOverflow((cluster, node) -> {
+            if (node == 1) // only node1 is upgraded, and the cluster is in mixed versions mode
+            {
+                assertPolicyTriggersAt2038(cluster.coordinator(1));
+                assertPolicyTriggersAt2038(cluster.coordinator(2));
+            }
+            else // both nodes have been upgraded, and the cluster isn't in mixed version mode anymore
+            {
+                assertPolicyTriggersAt2038(cluster.coordinator(1));
+                assertPolicyTriggersAt2038(cluster.coordinator(2));
+
+                // We restart one node on 5.0 >nc hence 2038 should still be the limit as the other node is 5.0 <=nc
+                // We're on compatibility mode where nc and oa nodes are a possibility
+                cluster.get(1).shutdown().get();
+                cluster.get(1).config().set("storage_compatibility_mode", UPGRADING.toString());
+                cluster.get(1).startup();
+                assertPolicyTriggersAt2038(cluster.coordinator(1));
+                assertPolicyTriggersAt2038(cluster.coordinator(2));
+
+                // We restart the other node so they're all on 5.0 >nc hence 2106 should be the limit
+                cluster.get(2).shutdown().get();
+                cluster.get(2).config().set("storage_compatibility_mode", UPGRADING.toString());
+                cluster.get(2).startup();
+                assertPolicyTriggersAt2106(cluster.coordinator(1));
+                assertPolicyTriggersAt2106(cluster.coordinator(2));
+
+                // We restart the cluster out of compatibility mode once everything is 5.0oa TTL 2106
+                cluster.get(1).shutdown().get();
+                cluster.get(1).config().set("storage_compatibility_mode", NONE.toString());
+                cluster.get(1).startup();
+                cluster.get(2).shutdown().get();
+                cluster.get(2).config().set("storage_compatibility_mode", NONE.toString());
+                cluster.get(2).startup();
+                assertPolicyTriggersAt2106(cluster.coordinator(1));
+                assertPolicyTriggersAt2106(cluster.coordinator(2));
+            }
+        });
+    }
+
+    private static void testTTLOverflow(RunOnClusterAndNode runAfterNodeUpgrade) throws Throwable
+    {
+        new TestCase()
+                .nodes(2)
+                .nodesToUpgradeOrdered(1, 2)
+                .upgradesToCurrentFrom(v41)
```

Comment:
Strictly this should be at least `v40`, which also includes `v41`. There is a risk of OOM when doubling the number of tests, though. If that happens this can be split in two classes, inheriting from a common superclass.

Reply:
Done